packages/flutter_tools/lib/src/web/compiler_config.dart

```diff
@@ -0,0 +1,137 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+abstract class WebCompilerConfig {
+  const WebCompilerConfig();
+
+  /// Returns `true` if `this` represents configuration for the Wasm compiler.
+  ///
+  /// Otherwise, `false`–represents the JavaScript compiler.
+  bool get isWasm;
+
+  Map<String, String> toBuildSystemEnvironment();
+}
+
+/// Configuration for the Dart-to-Javascript compiler (dart2js).
+class JsCompilerConfig extends WebCompilerConfig {
+  const JsCompilerConfig({
+    required this.csp,
+    required this.dumpInfo,
+    required this.nativeNullAssertions,
+    required this.optimizationLevel,
+    required this.noFrequencyBasedMinification,
+    required this.sourceMaps,
+  });
+
+  const JsCompilerConfig.run({required bool nativeNullAssertions})
```

Comment:
is this a constructor? i'm also unfamiliar with the `this()` syntax, is this syntactic sugar over `this.csp = false, this.dumpInfo = false, ...`?

Reply:
Yes and yes.

Reply:
can you add a dartdoc describing when to call this constructor (it's not obvious from the name)